services/messaging/src/main/java/org/eclipse/hono/messaging/ForwardingDownstreamAdapter.java

```diff
@@ -327,6 +327,15 @@ public final void onClientAttach(final UpstreamReceiver client, final Handler<As
             createSender(targetAddress, replenishedSender -> handleFlow(replenishedSender, client)).compose(createdSender -> {
                 addSender(client, createdSender);
                 tracker.complete();
+                createdSender.detachHandler(detachHandler -> {
+                    if (detachHandler.succeeded()) {
+                        logger.info("downstream sender was closed detached [con: {}, link: {}]", client.getConnectionId(), client.getLinkId());
+                    } else {
+                        logger.warn("downstream sender was closed detached [con: {}, link: {}]: {}", client.getConnectionId(), client.getLinkId(), detachHandler.cause().getMessage());
+                    }
+                    createdSender.close();
+                    resultHandler.handle(Future.failedFuture(detachHandler.cause()));
```

Comment:
this seems to be wrong. the resultHandler has already been completed by the *tracker* Future. 
when the peer (the router) detaches this sender link, then we need to do some house keeping:
1. close the link and remove it from the list of downstream senders
2. close the upstream receiver link

Reply:
I also would prefer if we could pass in a close handler to the `SenderFactory.createSender()` method so that the set up of the link is done in one place only. WDYT?